citra-qt: Add customizable speed limit target

jroweboy opened the pull request:
Allows the user to customize the speed limit instead of a binary off/on frame limiter. 

Things to know:
* Adjusts the max frametime lag to 250ms instead to allow lower frame limits. If the max lag is too small, the clamp will prevent limiting to smaller values. In practice: at the original 25ms value, the limiter could clamp around values of 60%, so half speed play wasn't possible. 250ms was selected as it allows clamps around 6%. This can be changed if someone thinks that limiting to such values would be pointless.
* Adds a hotkey (+ / - by default) to adjust the speed limit by +- 5%
* Status bar changed to display the limit
![Status bar changed to display the limit](https://user-images.githubusercontent.com/952515/34682780-3045e546-f45d-11e7-8f7e-7ed6d0c8aea4.png)
* Using a spinbox lets the user choose a precise limit as its adjustable by +- 1%
![Using a spinbox lets the user choose a precise limit](https://user-images.githubusercontent.com/952515/34682845-687d8874-f45d-11e7-95fa-26468b50d669.png)

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/citra-emu/citra/3353)
<!-- Reviewable:end -->

neobrain commented:
Hi, this is neobot, using neobrain's account. I'm keeping an archive of versions of this PR:

2018-01-08T17:29:52Z: https://github.com/citra-emu/citra/compare/a66e4585a0e69cad572332a3f9c267b5cdb3424b...jroweboy:0148ea9cdee281a0f378ba0defb96b6edbcc297c

neobrain commented:
Hi, this is neobot, using neobrain's account. I'm keeping an archive of versions of this PR:

2018-01-08T17:39:18Z: https://github.com/citra-emu/citra/compare/a66e4585a0e69cad572332a3f9c267b5cdb3424b...jroweboy:a6fbd8c56900a843163154ecac772d535d428f89

neobrain commented:
Hi, this is neobot, using neobrain's account. I'm keeping an archive of versions of this PR:

2018-01-08T18:16:05Z: https://github.com/citra-emu/citra/compare/a66e4585a0e69cad572332a3f9c267b5cdb3424b...jroweboy:8379072658d673902413ca86063a22cd310ee3ff

neobrain commented:
Hi, this is neobot, using neobrain's account. I'm keeping an archive of versions of this PR:

2018-01-08T18:26:48Z: https://github.com/citra-emu/citra/compare/a66e4585a0e69cad572332a3f9c267b5cdb3424b...jroweboy:22be430238ae98ac671eda51a8a1bf4fac903c83

Dragios commented on src/citra_qt/configuration/configure_graphics.ui:
```diff
@@ -44,81 +44,98 @@
-          <property name="text">
-           <string>Limit framerate</string>
-          </property>
-         </widget>
```
The checkbox for limit framerate should remain. What if I wanna do benchmark and make speed test comparison. With the current setting it is not possible to do as I can't make it run at the highest speed possible since it is user defined.

jroweboy commented:
Set it to 0 and its disabled. (I didn't have a good idea about how best to
document this)

On Mon, Jan 8, 2018 at 10:53 PM, Dragios <notifications@github.com> wrote:

> *@Dragios* commented on this pull request.
> ------------------------------
>
> In src/citra_qt/configuration/configure_graphics.ui
> <https://github.com/citra-emu/citra/pull/3353#discussion_r160323761>:
>
> > @@ -44,81 +44,98 @@
>           </widget>
>          </item>
>          <item>
> -         <widget class="QCheckBox" name="toggle_framelimit">
> -          <property name="text">
> -           <string>Limit framerate</string>
> -          </property>
> -         </widget>
>
> The checkbox for limit framerate should remain. What if I wanna do
> benchmark and make speed test comparison. With the current setting it is
> not possible to do as I can't make it run at the highest speed possible
> since it is user defined.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/citra-emu/citra/pull/3353#pullrequestreview-87419091>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA6Iw5D63PUAYpYu8qbsTqAkrd_nYQXPks5tIv73gaJpZM4RWsre>
> .
>

jroweboy commented:
@wwylele i can't think of any place i've personally seen that, but sure, I'll add that in. Thanks!

wwylele commented on src/core/perf_stats.cpp:
```diff
@@ -77,22 +77,24 @@ double PerfStats::GetLastFrameTimeScale() {
     // Max lag caused by slow frames. Can be adjusted to compensate for too many slow frames. Higher
     // values increase the time needed to recover and limit framerate again after spikes.
-    constexpr microseconds MAX_LAG_TIME_US = 25ms;
+    constexpr microseconds MAX_LAG_TIME_US = 250ms;
```
I am assuming this number is changed for limiter value other than 100%. In this case should this number be a function of limiter value instead?

wwylele commented on src/core/perf_stats.cpp:
```diff
@@ -77,22 +77,24 @@ double PerfStats::GetLastFrameTimeScale() {
+    constexpr microseconds MAX_LAG_TIME_US = 250ms;
 
-    if (!Settings::values.toggle_framelimit) {
+    if (!Settings::values.frame_limit) {
```
`if (Settings::values.frame_limit == 0)`

wwylele commented on src/video_core/video_core.cpp:
```diff
@@ -20,7 +20,7 @@ std::unique_ptr<RendererBase> g_renderer; ///< Renderer plugin
 std::atomic<bool> g_hw_renderer_enabled;
 std::atomic<bool> g_shader_jit_enabled;
 std::atomic<bool> g_vsync_enabled;
-std::atomic<bool> g_toggle_framelimit_enabled;
```
Is this variable actually being used anywhere?

jroweboy replied:
Nope. I think it was originally meant to have it there for a frame limit a long time ago. Didn't really look too much into it, just saw it wasn't used so I removed it.

wwylele commented on src/core/perf_stats.cpp:
```diff
@@ -77,22 +77,24 @@ double PerfStats::GetLastFrameTimeScale() {
     auto now = Clock::now();
 
-    frame_limiting_delta_err += microseconds(current_system_time_us - previous_system_time_us);
+    double sleep_scale = (Settings::values.frame_limit / 100.0);
```
No need for parenthese.

jroweboy commented:
Thanks for the review! Changes since last update:
* Addressed review comments
* Made max lag time a function of target speed percent.
* Added a checkbox to enable/disable frame limiter
* Prevent frame_limit from under/overflowing
* Hide target speed percent when frame limiter is off

wwylele commented on src/citra_qt/configuration/configure_graphics.ui:
```diff
@@ -202,6 +222,16 @@
+     <x>20</x>
+     <y>20</y>
+    </hint>
+   </hints>
```
what does this "hint" block do?

jroweboy replied:
i... don't know actually. It was something added by qt designer, so i got curious and didn't find any information about it out there. i'll just remove it for now

wwylele commented on src/video_core/video_core.cpp:
```diff
@@ -20,7 +20,7 @@ std::unique_ptr<RendererBase> g_renderer; ///< Renderer plugin
 std::atomic<bool> g_shader_jit_enabled;
 std::atomic<bool> g_vsync_enabled;
-std::atomic<bool> g_toggle_framelimit_enabled;
+std::atomic<u16> g_frame_limit;
```
I mean, this is not used anywhere, either, right?

jroweboy commented:
Updated UI picture: 
![citra-qt_2018-01-23_23-16-18](https://user-images.githubusercontent.com/952515/35317334-98738a50-0093-11e8-9578-57196f0d24eb.png)

When its disabled, it will hide the `/ TARGET%` in the status bar as well

@B3n30 if you have a chance to review and merge this please :)

Dragios commented:
By right when uncheck "Limit speed percent" the spin box should be greyed out and changes are locked.

B3n30 commented:
Reviewed 4 of 10 files at r1, 1 of 1 files at r2, 6 of 7 files at r5, 1 of 1 files at r7, 4 of 4 files at r8.
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---

*[src/citra_qt/main.cpp, line 294 at r1](https://reviewable.io:443/reviews/citra-emu/citra/3353#-L3bVUj5vUR_ipfHW53q:-L3bVUj5vUR_ipfHW53r:b-o6kfk0) ([raw file](https://github.com/citra-emu/citra/blob/0148ea9cdee281a0f378ba0defb96b6edbcc297c/src/citra_qt/main.cpp#L294)):*
> ```cpp
>     connect(GetHotkey("Main Window", "Increase Speed Limit", this), &QShortcut::activated, this,
>             [&] {
>                 Settings::values.frame_limit += 5;
> ```

Not sure about this: Should these hotkeys only be enabled if the framelimiter is enabled?

---

*[src/citra_qt/configuration/configure_graphics.cpp, line 99 at r8](https://reviewable.io:443/reviews/citra-emu/citra/3353#-L3bXeloXr1pf-SWuTyp:-L3bXeloXr1pf-SWuTyq:b-32g1bc) ([raw file](https://github.com/citra-emu/citra/blob/3106bfbe469e4833f3621df892744f63836bb65b/src/citra_qt/configuration/configure_graphics.cpp#L99)):*
> ```cpp
>         static_cast<int>(FromResolutionFactor(Settings::values.resolution_factor)));
>     ui->toggle_vsync->setChecked(Settings::values.use_vsync);
>     ui->toggle_frame_limit->setChecked(Settings::values.use_frame_limit);
> ```

Rename the check box to "use_frame_limit"?

---

*[src/core/telemetry_session.cpp, line 162 at r8](https://reviewable.io:443/reviews/citra-emu/citra/3353#-L3bYAZxwSb0Ca-omSFK:-L3bYAZxwSb0Ca-omSFL:b-w6oxo8) ([raw file](https://github.com/citra-emu/citra/blob/3106bfbe469e4833f3621df892744f63836bb65b/src/core/telemetry_session.cpp#L162)):*
> ```cpp
>     AddField(Telemetry::FieldType::UserConfig, "Renderer_ResolutionFactor",
>              Settings::values.resolution_factor);
>     AddField(Telemetry::FieldType::UserConfig, "Renderer_FrameLimit", Settings::values.frame_limit);
> ```

I think the use_frame_limit should still stay in telemetry

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/citra-emu/citra/3353)*
<!-- Sent from Reviewable.io -->

jroweboy commented:
Disabled the spin box when the checkbox is false

---

Review status: 10 of 12 files reviewed at latest revision, 8 unresolved discussions.

---

*[src/citra_qt/main.cpp, line 294 at r1](https://reviewable.io:443/reviews/citra-emu/citra/3353#-L3bVUj5vUR_ipfHW53q:-L3dIpFXKh19m5bdzz1G:bre6d6a) ([raw file](https://github.com/citra-emu/citra/blob/0148ea9cdee281a0f378ba0defb96b6edbcc297c/src/citra_qt/main.cpp#L294)):*
<details><summary><i>Previously, B3n30 wrote…</i></summary><blockquote>

Not sure about this: Should these hotkeys only be enabled if the framelimiter is enabled?
</blockquote></details>

I honestly don't think it matters at all, as if the frame limit is off, no one will even see a difference.

---

*[src/citra_qt/configuration/configure_graphics.cpp, line 99 at r8](https://reviewable.io:443/reviews/citra-emu/citra/3353#-L3bXeloXr1pf-SWuTyp:-L3dHtqMFcJqQ0aue2pY:b-qt2frr) ([raw file](https://github.com/citra-emu/citra/blob/3106bfbe469e4833f3621df892744f63836bb65b/src/citra_qt/configuration/configure_graphics.cpp#L99)):*
<details><summary><i>Previously, B3n30 wrote…</i></summary><blockquote>

Rename the check box to "use_frame_limit"?
</blockquote></details>

checkboxs in the ui are all prefixed with toggle so i'm keeping it that way for consistency\

---

*[src/core/perf_stats.cpp, line 80 at r4](https://reviewable.io:443/reviews/citra-emu/citra/3353#-L2Lo_G4uW5AY3VTuu03-r4-80:-L3dI5g5yW8ifWu2b-8k:b-hwd77o) ([raw file](https://github.com/citra-emu/citra/blob/22be430238ae98ac671eda51a8a1bf4fac903c83/src/core/perf_stats.cpp#L80)):*
<details><summary><i>Previously, wwylele (Weiyi Wang) wrote…</i></summary><blockquote>

I am assuming this number is changed for limiter value other than 100%. In this case should this number be a function of limiter value instead?
</blockquote></details>

Done. The function isn't anything special, and may need adjusting if someone 

---

*[src/core/perf_stats.cpp, line 82 at r4](https://reviewable.io:443/reviews/citra-emu/citra/3353#-L2Lo_G4uW5AY3VTuu03-r4-82:-L3dI524kICFwsKfnHmx:b-896fix) ([raw file](https://github.com/citra-emu/citra/blob/22be430238ae98ac671eda51a8a1bf4fac903c83/src/core/perf_stats.cpp#L82)):*
<details><summary><i>Previously, wwylele (Weiyi Wang) wrote…</i></summary><blockquote>

`if (Settings::values.frame_limit == 0)`
</blockquote></details>

Done.

---

*[src/core/perf_stats.cpp, line 88 at r4](https://reviewable.io:443/reviews/citra-emu/citra/3353#-L2Lo_G4uW5AY3VTuu03-r4-88:-L3dI6fLLIyIZqxeFNbz:b-896fix) ([raw file](https://github.com/citra-emu/citra/blob/22be430238ae98ac671eda51a8a1bf4fac903c83/src/core/perf_stats.cpp#L88)):*
<details><summary><i>Previously, wwylele (Weiyi Wang) wrote…</i></summary><blockquote>

No need for parenthese.
</blockquote></details>

Done.

---

*[src/core/telemetry_session.cpp, line 162 at r8](https://reviewable.io:443/reviews/citra-emu/citra/3353#-L3bYAZxwSb0Ca-omSFK:-L3dJQAhTsnokyIDHyYe:b-896fix) ([raw file](https://github.com/citra-emu/citra/blob/3106bfbe469e4833f3621df892744f63836bb65b/src/core/telemetry_session.cpp#L162)):*
<details><summary><i>Previously, B3n30 wrote…</i></summary><blockquote>

I think the use_frame_limit should still stay in telemetry
</blockquote></details>

Done.

---

*[src/video_core/video_core.cpp, line 23 at r7](https://reviewable.io:443/reviews/citra-emu/citra/3353#-L2Lo_G4uW5AY3VTuu07-r7-23:-L3dIDnaFX3B9XHma32k:b-896fix) ([raw file](https://github.com/citra-emu/citra/blob/46ea71f0de858cfeec905f0c9376cb22be1713bc/src/video_core/video_core.cpp#L23)):*
<details><summary><i>Previously, wwylele (Weiyi Wang) wrote…</i></summary><blockquote>

I mean, this is not used anywhere, either, right?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/citra-emu/citra/3353#-:-L3dVHPiT0bLFrM00Q0L:bjyj4oq)*
<!-- Sent from Reviewable.io -->

B3n30 commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 2 of 2 files at r9.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/citra-emu/citra/3353#-:-L3dYQFTRm3aVnUB6zeo:bnfp4nl)*
<!-- Sent from Reviewable.io -->

Dragios commented:
Fix clang format 1st before merging.

jroweboy commented:
I have a pr to update clang format to a new version and make it easier for
devs to use clang format coming up soon (after this is merged), so I don't
think it's that important. (Already got it merged to yuzu so it works just
fine as far as I've heard)

On Jan 24, 2018 6:52 PM, "Dragios" <notifications@github.com> wrote:

> Fix clang format 1st before merging.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/citra-emu/citra/pull/3353#issuecomment-360335313>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AA6Iw0V8l_kz0zi3vaY6PLt84czzo5tWks5tN95dgaJpZM4RWsre>
> .
>
